### DISCUSSION TITLE
How can I avoid the pixelation of ImageBrush?

### DISCUSSION THREAD

**Discussion by dbriard** [AvaloniaUI/Avalonia#11157] [Q&A]
Hi Avalonia Team and Community!

I am currently developing a design application that allow to manipulate visual elements, and I encounter a problem while converting WPF code to Avalonia. Specifically, the behavior of the ImageBrush is different between the two frameworks.

Suppose I have a view with a 100x100 pixel rectangle containing an ImageBrush of an image that is 1000x1000 pixels. I am facing two issues:

1) When zooming in by a factor of 10 in the view (using RenderTransform of the canva that contains the elements), WPF will display all 1000x1000 pixels of the image in the rectangle with the brush, while the same code in Avalonia will only show a texture of 100x100 pixel.

2) I also rely on BrushTransform to crop my images. When I crop 10% of the image (using a scale of x10 in the BrushTransform), WPF will display all the pixels of the image, while Avalonia scales the brush without utilizing extra resolution.

Is there a way to make the Avalonia ImageBrush behave like the WPF version?

Small element with large image zoomed in:
![image](https://user-images.githubusercontent.com/3595443/234857229-bd315dd3-0457-4e87-8282-22b92ec8c701.png)

**Comment by Gillibald** [comment]
Please share some code and also tell us what version of Avalonia you are using

**Reply by kekekeks** [reply]
We might need to expose UseScalableRasterization and make it to work with ImageBrush too.

**Reply by dbriard** [reply]
@Gillibald @kekekeks 

Thank you, here is the code and the rendering in both Avalonia and WPF:
```
<Canvas RenderTransformOrigin="0,0">
        <Canvas.RenderTransform>
            <ScaleTransform ScaleX="10" ScaleY="10" />
        </Canvas.RenderTransform>
        <Rectangle Width="64" Height="48">
            <Rectangle.Fill>
                <ImageBrush Source="/maple-leaf-888807_640.jpg" />
            </Rectangle.Fill>
        </Rectangle>
    </Canvas>
```

![image](https://user-images.githubusercontent.com/3595443/234876358-ce899f7e-f9a1-4d8e-b68d-54d6a111d7f7.png)

And for the brush transform:

 ```
<Canvas>
        <Rectangle Width="64" Height="48">
            <Rectangle.Fill>
                <ImageBrush ImageSource="/maple-leaf-888807_640.jpg">
                    <ImageBrush.Transform>
                        <ScaleTransform ScaleX="10" ScaleY="10"/>
                    </ImageBrush.Transform>
                </ImageBrush>
            </Rectangle.Fill>
        </Rectangle>
    </Canvas>
```

![image](https://user-images.githubusercontent.com/3595443/234877428-5a6a7306-7e5b-4f1e-8f4d-fb31eb231056.png)

I can share the projects if you need?

Of the two issues, the brush transform behavior is more problematic to me, as I am using it to crop inside a texture.
About zooming, in the view and not seeing extra pixel... it is disturbing as it do not work that way in WPF or UWP, but I can understand that it zoom what it see... but an option such a UseScalableRasterization would be welcome :)

It is another topic, but I think that would be useful to have a RelativeTransform property in Brush which work in (0,1) coordinates (as WPF have). I can live without it, and can do math and conversion, but it can be more user friendly to use: you could define a subpart of the image without knowing the size of the element for example, and it will still work when the element size change.

**Reply by dbriard** [reply]
> We might need to expose UseScalableRasterization and make it to work with ImageBrush too.

Hi @kekekeks ,

Is there a chance this could be done before v11 release? Should I open an issue for that?

Is there something I can try in the meantime? such as using Reflection to change UseScalableRasterization property somewhere? 

I've see there is a DestinationRect property that can be used for cropping, but as I also need crop-rotation, I can't use it.

Alternatively, I suppose I can replace my rectangle with transformed brush with a custom control and draw the image in Render with the desired transform applied to the drawingcontext. Do you think I'll have the same issue? and also, I am not sure about performance compared to image brush...

Thank you.

**Reply by kekekeks** [reply]
>Is there a chance this could be done before v11 release? Should I open an issue for that?

At the current stage we are focusing on general stability and adding the remaining breaking changes for the public API surface, so it would likely have to wait until some update release later this year.

>Is there something I can try in the meantime?
>Alternatively, I suppose I can replace my rectangle with transformed brush with a custom control and draw the image in Render with the desired transform applied to the drawingcontext. Do you think I'll have the same issue? and also, I am not sure about performance compared to image brush...


Right now "scalable rasterization" (that uses SKPicture instead of a materialized intermediate texture) is only used for DrawingBrush with a potential option to enable it for VisualBrush. ImageBrush goes through a different codepath, so it's not a simple switch.

You can try using DrawingBrush with ImageDrawing inside and check if that works for you
